### ISSUE TITLE
ConsoleEncoder: add option to format the context as key=value pairs

### ISSUE THREAD

**Issue by dnephin** [uber-go/zap#1076]
This is my first time using zap. I've used a couple other logging libraries in Go ([logrus](https://github.com/sirupsen/logrus), [go-hclog](https://github.com/hashicorp/go-hclog)). In both of those libraries the "human readable" format would format the context as `key=value` pairs. I was a bit surprised to see that zap encodes the whole context as JSON.

What I'd like is to use the `ConsoleEncoder` when the application is run with a terminal (dev and test), and switch to the `JSONEncoder` when the application is run in production.  That mostly works, but a lot of the important details in the log are part of the context, and formatting the context as JSON makes it not super readable.

Do you think it would be reasonable to add an option to the `ConsoleEncoder` to allow formatting the context as `key=value` pairs instead of JSON? Would you accept a PR that adds that support?

Thank you!

**Comment by abhinav** [comment]
tl;dr: Skip to the bottom

Hey @dnephin, that's a valid point.  I think the output of console encoder could bear to be more readable, and we're definitely open to the idea of changing it to that effect. Although, I'm not sure whether this should be an option on the existing console encoder or a completely new console encoder.

To elaborate on my thought process there: consider that first we need to decide whether we want super specific options like "key value delimiter" (e.g. `=` or `:`) and "object field delimiter" (e.g. `,` or `\t`) or a more general option for "be more readable" where we're allowed to change the output and make it more readable—whatever we consider that to be. I lean towards the latter: users choose between machine-readable and human-readable output, not the specifics of what "human-readable" is. And that's exactly what the separation between JSONEncoder and ConsoleEncoder is supposed to represent. So _theoretically_, we can just change the output of the existing console encoder. However, I'm concerned that it would break systems like [zap-pretty](https://github.com/maoueh/zap-pretty) that consume this output. (Well, not zap-pretty specifically—it consumes the JSON encoder's output—but systems like it.) We could make it opt-in, but that means that the default behavior will be less readable.

...which is what brought up my original question of option versus new encoder. A new encoder would be allowed to be "readable by default", and could establish, as part of its contract, that its output may change to become more readable in the future.

Personally, I'm okay with changing the console encoder to be more readable because I think "be more readable" is it's contract, but I want to solicit more opinions. I'll reach out to some folks. I'd also like to hear from the community.

---

**tl;dr: Should the console encoder be allowed to change its output if we think it'll be more readable?**

**Comment by mway** [comment]
I think changing the console encoder's default output is okay:

- Consuming log entries should be done either via a proper serialization format (e.g. a [JSON Encoder](https://pkg.go.dev/go.uber.org/zap/zapcore#NewJSONEncoder)) or via API (e.g. [zaptest/observer](https://pkg.go.dev/go.uber.org/zap/zaptest/observer)), because those have actual contracts. Mixed-mode data (e.g. output from the console encoder) shouldn't be used for programmatic purposes.
- Consoles are primarily for humans, and humans are more adaptable/flexible than machines.

If we think it's valuable, we could add an opt-in option to restore the previous behavior (e.g. a new `HumanizeContext` option that defaults to true, but users could set to false as needed).

**Comment by sagarsuperuser** [comment]
Can someone please guide me on this ->
https://github.com/uber-go/zap/issues/1509
I guess I have created the solution (a TextEncoder) for this same issue.
Added benchmarks, tests and examples reference in the issue https://github.com/uber-go/zap/issues/1509
